sql/core/src/main/scala/org/apache/spark/sql/execution/ExistingRDD.scala

```diff
@@ -217,19 +241,29 @@ private[sql] case class DataSourceScan(
     // TODO: The abstractions between this class and SqlNewHadoopRDD makes it difficult to know
     // here which path to use. Fix this.
 
-    ctx.INPUT_ROW = row
+    // These assignments avoids accesses to instance variables in a while-loop
+    // Since they are loop invariant variables (LIVs), host them outside the while-loop
+    val colLocalVars = output.zipWithIndex.map(x => ctx.freshName("col" + x._2))
+    val columnLIVAssigns = (colLocalVars zip colVars).map { case (localName, name) =>
+      s"$columnVectorClz $localName = $name;" }
+
     ctx.currentVars = null
-    val columns1 = exprs.map(_.gen(ctx))
+    val columns1 = (output zip colLocalVars).map { case (attr, colVar) =>
+      genCodeColumnVector(ctx, colVar, rowidx, attr.dataType, attr.nullable) }
     val scanBatches = ctx.freshName("processBatches")
     ctx.addNewFunction(scanBatches,
       s"""
       | private void $scanBatches() throws java.io.IOException {
       |  while (true) {
       |     int numRows = $batch.numRows();
-      |     if ($idx == 0) $numOutputRows.add(numRows);
+      |     if ($idx == 0) {
+      |       ${columnAssigns.mkString("", "\n", "\n")}
+      |       $numOutputRows.add(numRows);
+      |     }
       |
+      |     ${columnLIVAssigns.mkString("", "\n", "\n")}
```

Comment:
For now, I will not use this. In the future, I may revisit this scala replacement regarding all of possible variables in another PR.